resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala

```diff
@@ -381,4 +382,21 @@ object KubernetesUtils extends Logging {
       }
     }
   }
+
+  @Since("3.3.0")
+  def createOrReplaceResource(
+      client: KubernetesClient,
+      resources: Seq[HasMetadata],
+      ownerPod: Pod = null): Unit = {
+    try {
+      addOwnerReference(ownerPod, resources)
+      client.resourceList(resources: _*).createOrReplace()
+    } catch {
+      case NonFatal(e) =>
+        if (ownerPod != null) {
+          client.pods().delete(ownerPod)
```

Comment:
Well, this logic looks unsafe because the pre-populated resources will be leaked if the failure happens at line 392, `addOwnerReference(ownerPod, resources)`.